lib/defender/aws_iot_defender_mqtt.c

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (C) 2018 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/* Defender internal include. */
+#include "private/aws_iot_defender_internal.h"
+
+/*
+ * 30 seconds is the minimum allowed in AFR MQTT
+ */
+#define _DEFENDER_MQTT_KEEP_ALIVE_SECONDS             30
+
+/* Define topics segments used by defender. */
+#define _TOPIC_PREFIX             "$aws/things/"
+
+#define _TOPIC_SUFFIX_PUBLISH     "/defender/metrics/" _DEFENDER_FORMAT
+
+#define _TOPIC_SUFFIX_ACCEPTED    _TOPIC_SUFFIX_PUBLISH "/accepted"
+
+#define _TOPIC_SUFFIX_REJECTED    _TOPIC_SUFFIX_PUBLISH "/rejected"
+
+/* defender internally manages network and mqtt connection */
+static AwsIotNetworkConnection_t _networkConnection = AWS_IOT_NETWORK_CONNECTION_INITIALIZER;
+static AwsIotMqttConnection_t _mqttConnection = AWS_IOT_MQTT_CONNECTION_INITIALIZER;
+
+static char * _pPublishTopic = NULL;
+
+static char * _pAcceptTopic = NULL;
+
+static char * _pRejectTopic = NULL;
+
+/*-----------------------------------------------------------*/
+
+bool AwsIotDefenderInternal_BuildTopicsNames( const char * pThingName,
+                                              uint16_t thingNameLength,
+                                              AwsIotDefenderError_t * error )
+{
+    /* calculate topics lengths. */
+    size_t publishTopicLength = strlen( _TOPIC_PREFIX ) + thingNameLength + strlen( _TOPIC_SUFFIX_PUBLISH ) + 1;
+    size_t acceptTopicLength = strlen( _TOPIC_PREFIX ) + thingNameLength + strlen( _TOPIC_SUFFIX_ACCEPTED ) + 1;
+    size_t rejectTopicLength = strlen( _TOPIC_PREFIX ) + thingNameLength + strlen( _TOPIC_SUFFIX_REJECTED ) + 1;
+
+    /* allocate memory for each of them. */
+    char * pPublishTopic = pvPortMalloc( publishTopicLength * sizeof( char ) );
```

Comment:
This library should be portable to C-SDK, so you need to use an abstractions. Our model with MQTT is currently to define a macro like:
#define AwsIotMqtt_MallocOperation malloc
...